// node-hypixel/index.js
const request = require('request');
const querystring = require('querystring');
const base = 'https://api.hypixel.net';

class hypixel {
    constructor(key) {
        this.key = key;
    }
    
    buildPath(path, query) {
        const params = query || null;

        const _query = querystring.stringify(Object.assign({}, params, {
            key: this.key,
        }));

        return `${base}/${path}?${_query}`;
    }

    sendRequest2(path, query, resultField, callback) {
        request(this.buildPath(path, query), (error, res, body) => {
            let data = null;

            if (!error) {
                try {
                    data = JSON.parse(body);
                } catch (ex) {
                    return callback(new Error('malformed json'), null);
                }
            }

            if (data) {
                if (data.success) {
                    return callback(error, resultField ? data[resultField] : data);
                }

                return callback(new Error(data.cause), null);
            }

            return callback(error, data);
        });
    }

    sendRequest(path, query, resultField, callback) {
        if (callback) {
            return this.sendRequest2(path, query, resultField, callback);
        }

        
        return new Promise((resolve, reject) => {
            this.sendRequest2(path, query, resultField, (error, data) => {
                if (error) {
                    reject(error);
                } else {
                    resolve(data);
                }
            });
        });
    }

    getKeyInfo(callback) {
        return this.sendRequest('key', null, 'record', callback);
    }

    findGuild(field, value, callback) {
        return this.sendRequest('findGuild', { [field]: value }, 'guild', callback);
    }
    findGuildByName(name, callback) {
        return this.findGuild('byName', name, callback);
    }
    findGuildByPlayer(player, callback) {
        return this.findGuild('byUuid', player, callback);
    }
    getGuild(id, callback) {
        return this.sendRequest('guild', { id }, 'guild', callback);
    }
    getBoosters(callback) {
        return this.sendRequest('boosters', null, 'boosters', callback);
    }
    getFriends(player, callback) {
        return this.sendRequest('friends', { uuid: player }, 'records', callback);
    }
    getPlayer2(field, value, callback) {
        return this.sendRequest('player', { [field]: value }, 'player', callback);
    }
    getPlayer(id, callback) {
        return this.getPlayer2('uuid', id, callback);
    }
    getPlayerByUsername(username, callback) {
        return this.getPlayer2('name', username, callback);
    }
    getResources(resourceType, callback) {
        resourceType = resourceType || "achievements"
        return this.sendRequest("resources/" + resourceType, null, null, callback)
    }
    getGamesCount(callback) {
        return this.sendRequest("gameCounts", null, null, callback)
    }
    getPlayersCount(callback) {
        return this.sendRequest("playerCount", null, null, callback)
    }
    getRecent(uuid, callback) {
        return this.sendRequest("recentGames", {uuid: uuid}, "games", callback)
    }
    getStatus(uuid, callback) {
        return this.sendRequest("status", {uuid: uuid}, "session", callback)
    }
    getLeaderboard(callback) {
        return this.sendRequest("leaderboards", null, null, callback)
    }
    getWatchdogStats(callback) {
        return this.sendRequest("watchdogstats", null, null, callback)
    }
}

module.exports = hypixel
